### PR TITLE
Refactor cmd/nerdctl

### PR DESCRIPTION
**Pull request by apostasie**
This is in light of #3375 and #1680

The overall goal is to break down cmd/nerdctl.

These commits do:
- move some commonly used functions into `cmd/nerdctl/helpers`
  - these have been roughly split into distinct files that belong to different categories (testing, cobra, etc.)
  - see final note below on not attempting to clean these up further for now
- move to subpackages:
  - `cmd/nerdctl/completion`
  - `cmd/nerdctl/login`
  - `cmd/nerdctl/namespace`
  - `cmd/nerdctl/network`
  - `cmd/nerdctl/volume`
  - `cmd/nerdctl/image`
  - `cmd/nerdctl/builder`
  - `cmd/nerdctl/apparmor`
  - `cmd/nerdctl/internal`

This is purely refactoring, and there is no functional or logic change (see testing note at the bottom though).

Further note that this PR does not try to "cleanup" anything besides that. Specifically out of scope for now (could be done in a subsequent PR):
- name stuttering reduction
- possible cleanup of the "helpers" folder

Also, this PR is big enough now - what remains to be done (containers and compose) are big chunks on their own, and I would rather merge this first then tackle these two in the next PR.

This is on-top of #3384 that should be merged first.

### CODE REVIEW

**Comment by apostasie** [comment]
@AkihiroSuda @djdongjin and others: lmk if you think this is going in the right direction (before I invest more effort into this).

File: cmd/nerdctl/completion/completion.go [hunk 1]
```diff
@@ -24,13 +24,17 @@ import (
...
 	"github.com/containerd/nerdctl/v2/pkg/clientutil"
+	"github.com/containerd/nerdctl/v2/pkg/cmd/volume"
+	"github.com/containerd/nerdctl/v2/pkg/inspecttypes/native"
 	"github.com/containerd/nerdctl/v2/pkg/labels"
 	"github.com/containerd/nerdctl/v2/pkg/netutil"
 )
 
-func shellCompleteImageNames(cmd *cobra.Command) ([]string, cobra.ShellCompDirective) {
-	globalOptions, err := processRootCmdFlags(cmd)
+func ShellCompleteImageNames(cmd *cobra.Command) ([]string, cobra.ShellCompDirective) {
```
Comment (AkihiroSuda) [inline cmd/nerdctl/completion/completion.go:36]:
```suggestion
func ImageNames(cmd *cobra.Command) ([]string, cobra.ShellCompDirective) {
```

**Review by AkihiroSuda — APPROVED** [review]
Thanks, I think this direction is fine, as long as the CI can remain green

File: cmd/nerdctl/completion/completion.go (hunk 1 above)
Comment (apostasie) [reply cmd/nerdctl/completion/completion.go]: Done for all completion functions.

**Comment by apostasie** [comment]
@Zheaoli lmk what you think.
Otherwise, happy to have this first one merged at your convenience folks.
Will move forward in that direction then.

**Comment by apostasie** [comment]
@AkihiroSuda I am sending atomic commits on this, where each commit does one simple/clear thing.
Hopefully, that should help reviewers - as my latest PRs have been a big too big/compact and people have been asking me for more broken down, easier to review chunks.

In the case here, though, lmk if you folks would prefer me to squash.

**Comment by AkihiroSuda** [comment]
Thanks, in this case commits should remain unsquashed :+1:

**Comment by apostasie** [comment]
@AkihiroSuda I clearly hit a snag wrt testing (maybe that is what stopped the previous refactoring attempt).
Will investigate and fix (but reverting to draft until the CI is green).

This is still good to review right now - at least to validate that in principle we are all fine with the direction and that "what was moved where" does make sense.

**Comment by Zheaoli** [comment]
> @Zheaoli lmk what you think. Otherwise, happy to have this first one merged at your convenience folks. Will move forward in that direction then.

This PR is mostly LGTM. BTW I think maybe we can split it into multiple some PR.

**Comment by apostasie** [comment]
> > @Zheaoli lmk what you think. Otherwise, happy to have this first one merged at your convenience folks. Will move forward in that direction then.
> 
> This PR is mostly LGTM. BTW I think maybe we can split it into multiple some PR.

Thanks a lot @Zheaoli !
Appreciate your review.

I did split this into a bunch of separate commits - so, that should be easing review in the same way different PRs would.

That being said, I also split it into two PRs (next one being: #3383) as it was indeed growing out of control.

**Comment by apostasie** [comment]
@sondavidb 

This PR **last commit** breaks windows: 959567906457fa76655efaeef4fce73ce783baf5

It does work fine before this commit.

The test that breaks down is `TestImageInspectContainsSomeStuff` (great test name), but this might be a red herring.

The failure message is Klingon to me: "failed to extract layer sha256:f2032cdb21dfb9f3d6fdd3797ea992ec05415e5a6a57f9eb36b6646496db8d09: hcsshim::ProcessUtilityVMImage \\\\?\\C:\\ProgramData\\containerd\\root\\io.containerd.snapshotter.v1.windows\\snapshots\\2\\UtilityVM: Incorrect function.: unknown""

I have read this commit over and over, and cannot find anything blatantly wrong - my current intuition is that it breaks because the tests order changed, or because the test got relocated (before that commit, that test was ran under `cmd/nerdctl`, and after all the `containerx_test` tests).

So, either the location change has an impact, or something happening in containers tests that somehow "leak" into this test.

At this point, I cannot go anywhere without a windows rig, so, I will bite it and go get some windows rig... (or is it doable inside a VM on mac?)

But on the off-chance you could take a look, I thought I would ask you:
- is the error message any familiar?
- would you be able to play around with/without that commit and see if you get anywhere?

Thanks :-)

**Comment by apostasie** [comment]
@AkihiroSuda @Zheaoli CI is green.

This is ready AFAIC.

We need to merge #3384 first as it has important fixes for testing, and then this.

File: .github/workflows/test-canary.yml [hunk 2]
```diff
@@ -112,4 +112,5 @@ jobs:
           ctrdVersion: ${{ env.CONTAINERD_VERSION }}
         run: powershell hack/configure-windows-ci.ps1
       - name: "Run integration tests"
-        run: go test -v ./cmd/...
+        # See https://github.com/containerd/nerdctl/blob/main/docs/testing.md#about-parallelization
+        run: go test -p 1 -v ./cmd/nerdctl/...
```
Comment (AkihiroSuda) [inline .github/workflows/test-canary.yml:116]: Could you rebase the PR?

**Review by AkihiroSuda — APPROVED** [review]
Thanks

**Pull request merged**